Input
=====

*OmicScope* offers five methods for integrating data into its pipeline, three of which rely on proteomic software for protein identification and quantitative analyses:


* 
  **Progenesis QI for proteomics** *(Method = 'Progenesis')*\ : Progenesis is a software that enables protein quantification and identification (coupled to PLGS) for experiments that use Data Independent Acquisition (DIA).

* 
  **PatternLab V** *(Method = 'PatternLab')*\ : PatternLab V is an integrated computational environment for analyzing shotgun proteomic data, and is considered one of the best options for quantitative proteomics using data-dependent acquisition, due to its high-confidence parameters for protein quantitation and identification.

* 
  **MaxQuant** *(Method = 'MaxQuant')*\ : MaxQuant is the most widely used software for quantitative proteomics, offering users a range of parameter options for quantitative analysis.

* 
  **General** *(Method = 'General')*\ : A generic input method that requires users to specify quantitative values (sheet1 = assay), protein features (sheet2 = rdata), and sample information (sheet3 = pdata) using an Excel file.

* 
  **Snapshot** *(Method = 'Snapshot')*\ : The Snapshot method is the simplest approach to be used in the OmicScope workflow. It involves using a single, concise Excel sheet that contains essential information about proteins in a study, including fold change, p-value, and more.

Import OmicScope
----------------

.. code-block:: python

   import omicscope as omics

.. code-block::

   OmicScope v 1.0.5 For help: Insert
   If you use  in published research, please cite: 'XXXXX'
   Reis-de-Oliveira G, Martins-de-Souza D. OmicScope: from quantitative proteomics to systems biology.




Progenesis QI for Proteomics
----------------------------

Progenesis exports a .csv file containing information about samples, proteins, and quantitative values. For the Progenesis workflow, *OmicScope* imports these files and extracts the abundance levels of each protein (assay), the features of each protein (rdata), and features of each sample (pdata). Additionally, *OmicScope* accepts Excel files (with extensions .xls or .xlsx) that contain a **unique sheet** for the Progenesis workflow, as many users may use Excel to open and manipulate data. 

.. code-block:: python

   progenesis = omics.OmicScope('../tests/data/proteins/progenesis.xls', Method='Progenesis')

.. code-block::

   User already performed statistical analysis



In addition, since Progenesis exports a limited amount of information about samples, *OmicScope* allows the user to input an Excel file containing all this information. Furthermore, users can filter data based on the minimum number of unique peptides, by adding the parameter function 'UniquePeptides' (recommended: UniquePeptides = 1).

**Note:** As Progenesis performs differential proteomics, *OmicScope* takes into account the statistical analysis exported by Progenesis. However, if the user has a specific experimental design and requires the *OmicScope* Statistical Workflow, they must rewrite the original .csv file as follows:


* "Anova (p)" → "Original Anova (p)"
* "q Value" → "Original q Value"

.. code-block:: python

   progenesis_uniquepepfilt = omics.OmicScope('../tests/data/proteins/progenesis.xls', Method='Progenesis', UniquePeptides=1)
   print('Original proteomics data: ' + str(len(progenesis.quant_data)) + '\n'+
         'Filtered proteomics data: ' + str(len(progenesis_uniquepepfilt.quant_data))
         )

.. code-block::

   User already performed statistical analysis
   Original proteomics data: 2236
   Filtered proteomics data: 1847



PatternLab
----------

PatternLab exports an Excel file with an .xlsx extension, which contains the same type of information as Progenesis, including assay, pdata, and rdata. However, this export does not include differential proteomics statistics. Therefore, *OmicScope* automatically performs statistical analysis for PatternLab data.

.. code-block:: python

   plv = omics.OmicScope('../tests/data/proteins/patternlab.xlsx', Method='PatternLab')

.. code-block::

   Anova test was performed!
   OmicScope performed statistical analysis (Static workflow)



MaxQuant
--------

MaxQuant exports a **proteinGroups** file with a .txt extension, which contains a comprehensive description of the assay and rdata. However, since pdata is missing, the MaxQuant workflow **requires** an Excel file for pdata.

.. code-block:: python

   maxquant = omics.OmicScope('../tests/data/proteins/MQ.txt', Method='MaxQuant',
                              pdata='../tests/data/proteins/MQ_pdata.xlsx')

.. code-block::

   Anova test was performed!
   OmicScope performed statistical analysis (Static workflow)



General
-------

The General workflow allows users to analyze data generated by other platforms, such as Transcriptomics and Metabolomics. To do this, users need to organize an Excel file into three sheets containing an assay, rdata, and pdata.


* **Assay:** Contains the abundance of N proteins (rows) from M samples (columns).
* **Rdata:** Includes N proteins (rows) with their respective features within each column.
* **Pdata:** Contains M samples (rows) with their respective characteristics, such as conditions, biological and technical replicates.

The following sections provide examples of how to describe each sheet.

.. code-block:: python

   general = omics.OmicScope('../tests/data/proteins/general.xlsx', Method='General')

.. code-block::

   User already performed statistical analysis



Assay
^^^^^

.. code-block:: python

   import pandas as pd

   assay = pd.read_excel('../tests/data/proteins/general.xlsx', sheet_name=0)
   assay


.. raw:: html

   <div>
   <style scoped>
       .dataframe tbody tr th:only-of-type {
           vertical-align: middle;
       }

       .dataframe tbody tr th {
           vertical-align: top;
       }

       .dataframe thead th {
           text-align: right;
       }
   </style>
   <table border="1" class="dataframe">
     <thead>
       <tr style="text-align: right;">
         <th></th>
         <th>VCC_HB_1_1_2020</th>
         <th>VCC_HB_1_2</th>
         <th>VCC_HB_2_1</th>
         <th>VCC_HB_2_1_2</th>
         <th>VCC_HB_3_1</th>
         <th>VCC_HB_3_1_2</th>
         <th>VCC_HB_4_1</th>
         <th>VCC_HB_4_1_2</th>
         <th>VCC_HB_5_1</th>
         <th>VCC_HB_5_1_2</th>
         <th>...</th>
         <th>VCC_HB_C_1</th>
         <th>VCC_HB_C_1_2</th>
         <th>VCC_HB_D_1</th>
         <th>VCC_HB_D_1_2</th>
         <th>VCC_HB_E_1</th>
         <th>VCC_HB_E_1_2</th>
         <th>VCC_HB_F_1</th>
         <th>VCC_HB_F_1_2</th>
         <th>VCC_HB_G_1</th>
         <th>VCC_HB_G_1_2</th>
       </tr>
     </thead>
     <tbody>
       <tr>
         <th>0</th>
         <td>2.938847e+04</td>
         <td>3.110927e+04</td>
         <td>2.521807e+04</td>
         <td>3.090703e+04</td>
         <td>2.383499e+04</td>
         <td>2.267237e+04</td>
         <td>3.290661e+04</td>
         <td>3.595543e+04</td>
         <td>3.116039e+04</td>
         <td>3.083176e+04</td>
         <td>...</td>
         <td>1.318972e+04</td>
         <td>1.295723e+04</td>
         <td>1.438445e+04</td>
         <td>1.345175e+04</td>
         <td>1.136301e+04</td>
         <td>1.296883e+04</td>
         <td>1.273169e+04</td>
         <td>1.323385e+04</td>
         <td>1.505976e+04</td>
         <td>1.242351e+04</td>
       </tr>
       <tr>
         <th>1</th>
         <td>7.081308e+04</td>
         <td>6.446946e+04</td>
         <td>5.825493e+04</td>
         <td>5.931610e+04</td>
         <td>6.309095e+04</td>
         <td>5.933534e+04</td>
         <td>4.769155e+04</td>
         <td>4.923312e+04</td>
         <td>5.540644e+04</td>
         <td>5.816974e+04</td>
         <td>...</td>
         <td>1.177108e+04</td>
         <td>1.309192e+04</td>
         <td>1.790318e+04</td>
         <td>1.597053e+04</td>
         <td>1.299094e+04</td>
         <td>1.297193e+04</td>
         <td>1.732675e+04</td>
         <td>2.065285e+04</td>
         <td>1.496680e+04</td>
         <td>1.349288e+04</td>
       </tr>
       <tr>
         <th>2</th>
         <td>1.007536e+05</td>
         <td>1.011999e+05</td>
         <td>7.301329e+04</td>
         <td>7.349391e+04</td>
         <td>9.766835e+04</td>
         <td>9.952204e+04</td>
         <td>9.291146e+04</td>
         <td>7.996892e+04</td>
         <td>8.602800e+04</td>
         <td>8.490852e+04</td>
         <td>...</td>
         <td>1.224955e+05</td>
         <td>1.330394e+05</td>
         <td>1.223455e+05</td>
         <td>1.248028e+05</td>
         <td>1.051494e+05</td>
         <td>9.843347e+04</td>
         <td>1.224127e+05</td>
         <td>1.154907e+05</td>
         <td>1.364931e+05</td>
         <td>1.432545e+05</td>
       </tr>
       <tr>
         <th>3</th>
         <td>2.588031e+04</td>
         <td>3.769105e+04</td>
         <td>2.992691e+04</td>
         <td>3.460095e+04</td>
         <td>2.596320e+04</td>
         <td>2.578964e+04</td>
         <td>3.433567e+04</td>
         <td>3.174725e+04</td>
         <td>5.536738e+04</td>
         <td>6.139519e+04</td>
         <td>...</td>
         <td>2.827172e+04</td>
         <td>2.642189e+04</td>
         <td>2.325916e+04</td>
         <td>2.394876e+04</td>
         <td>2.288316e+04</td>
         <td>2.246984e+04</td>
         <td>1.679230e+04</td>
         <td>1.792154e+04</td>
         <td>2.125956e+04</td>
         <td>2.326553e+04</td>
       </tr>
       <tr>
         <th>4</th>
         <td>1.019192e+06</td>
         <td>1.109406e+06</td>
         <td>1.060396e+06</td>
         <td>1.078239e+06</td>
         <td>1.003426e+06</td>
         <td>9.853668e+05</td>
         <td>9.319196e+05</td>
         <td>8.415161e+05</td>
         <td>1.046464e+06</td>
         <td>1.094484e+06</td>
         <td>...</td>
         <td>9.032477e+05</td>
         <td>9.224474e+05</td>
         <td>8.579011e+05</td>
         <td>8.631613e+05</td>
         <td>8.902884e+05</td>
         <td>9.156146e+05</td>
         <td>7.709503e+05</td>
         <td>7.985187e+05</td>
         <td>8.478530e+05</td>
         <td>8.362974e+05</td>
       </tr>
       <tr>
         <th>...</th>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
       </tr>
       <tr>
         <th>2273</th>
         <td>3.654582e+06</td>
         <td>4.052283e+06</td>
         <td>3.865664e+06</td>
         <td>4.097018e+06</td>
         <td>3.904966e+06</td>
         <td>3.948974e+06</td>
         <td>3.361720e+06</td>
         <td>3.377376e+06</td>
         <td>3.553030e+06</td>
         <td>3.585600e+06</td>
         <td>...</td>
         <td>3.683729e+06</td>
         <td>3.937861e+06</td>
         <td>3.958128e+06</td>
         <td>3.966355e+06</td>
         <td>3.891880e+06</td>
         <td>3.949765e+06</td>
         <td>3.785332e+06</td>
         <td>3.987608e+06</td>
         <td>4.958008e+06</td>
         <td>5.095684e+06</td>
       </tr>
       <tr>
         <th>2274</th>
         <td>3.404704e+06</td>
         <td>3.868381e+06</td>
         <td>1.460918e+07</td>
         <td>1.770459e+07</td>
         <td>3.409452e+06</td>
         <td>3.743869e+06</td>
         <td>4.549866e+06</td>
         <td>4.549434e+06</td>
         <td>4.131214e+06</td>
         <td>4.325119e+06</td>
         <td>...</td>
         <td>2.855731e+06</td>
         <td>3.215217e+06</td>
         <td>4.188039e+06</td>
         <td>4.599402e+06</td>
         <td>4.887900e+06</td>
         <td>5.347565e+06</td>
         <td>4.919236e+06</td>
         <td>5.455081e+06</td>
         <td>4.757800e+06</td>
         <td>5.204147e+06</td>
       </tr>
       <tr>
         <th>2275</th>
         <td>2.249648e+05</td>
         <td>2.365908e+05</td>
         <td>2.035899e+05</td>
         <td>1.925923e+05</td>
         <td>2.638737e+05</td>
         <td>2.647051e+05</td>
         <td>3.204020e+05</td>
         <td>2.865846e+05</td>
         <td>2.102448e+05</td>
         <td>2.156030e+05</td>
         <td>...</td>
         <td>2.642441e+05</td>
         <td>2.660566e+05</td>
         <td>2.794582e+05</td>
         <td>2.686382e+05</td>
         <td>2.870855e+05</td>
         <td>2.940200e+05</td>
         <td>2.614172e+05</td>
         <td>2.533796e+05</td>
         <td>3.254633e+05</td>
         <td>3.297688e+05</td>
       </tr>
       <tr>
         <th>2276</th>
         <td>2.276456e+06</td>
         <td>2.976481e+06</td>
         <td>1.280703e+07</td>
         <td>1.644311e+07</td>
         <td>2.334264e+06</td>
         <td>2.590021e+06</td>
         <td>2.587180e+06</td>
         <td>2.693531e+06</td>
         <td>3.252054e+06</td>
         <td>3.787151e+06</td>
         <td>...</td>
         <td>1.739140e+06</td>
         <td>1.913206e+06</td>
         <td>2.827441e+06</td>
         <td>3.181103e+06</td>
         <td>3.306627e+06</td>
         <td>3.886518e+06</td>
         <td>3.709116e+06</td>
         <td>4.356867e+06</td>
         <td>3.156444e+06</td>
         <td>3.649251e+06</td>
       </tr>
       <tr>
         <th>2277</th>
         <td>1.575684e+05</td>
         <td>1.709010e+05</td>
         <td>2.185840e+05</td>
         <td>2.397707e+05</td>
         <td>1.621122e+05</td>
         <td>1.605849e+05</td>
         <td>6.805103e+04</td>
         <td>6.544271e+04</td>
         <td>1.724174e+05</td>
         <td>1.859680e+05</td>
         <td>...</td>
         <td>1.370555e+05</td>
         <td>1.445220e+05</td>
         <td>1.366811e+05</td>
         <td>1.409139e+05</td>
         <td>1.526514e+05</td>
         <td>1.631331e+05</td>
         <td>1.467560e+05</td>
         <td>1.543588e+05</td>
         <td>1.805445e+05</td>
         <td>1.884684e+05</td>
       </tr>
     </tbody>
   </table>
   <p>2278 rows × 38 columns</p>
   </div>


rdata
^^^^^

The rdata sheet needs to have at least two columns: 'Accession' and 'Description'.


#. **Accession:** An array of unique values that represent the proteins in the dataframe.
#. **Description:** The header from UniProt Fasta.

.. code-block:: python

   rdata = pd.read_excel('../tests/data/proteins/general.xlsx', sheet_name=1)
   rdata


.. raw:: html

   <div>
   <style scoped>
       .dataframe tbody tr th:only-of-type {
           vertical-align: middle;
       }

       .dataframe tbody tr th {
           vertical-align: top;
       }

       .dataframe thead th {
           text-align: right;
       }
   </style>
   <table border="1" class="dataframe">
     <thead>
       <tr style="text-align: right;">
         <th></th>
         <th>Accession</th>
         <th>Peptide count</th>
         <th>Unique peptides</th>
         <th>Confidence score</th>
         <th>Anova (p)</th>
         <th>q Value</th>
         <th>Max fold change</th>
         <th>Power</th>
         <th>Highest mean condition</th>
         <th>Lowest mean condition</th>
         <th>Description</th>
       </tr>
     </thead>
     <tbody>
       <tr>
         <th>0</th>
         <td>P0DJI8</td>
         <td>1</td>
         <td>1</td>
         <td>6.8809</td>
         <td>0.000000e+00</td>
         <td>0.000000</td>
         <td>2.192654</td>
         <td>1.000000</td>
         <td>COVID</td>
         <td>CTRL</td>
         <td>Serum amyloid A-1 protein OS=Homo sapiens OX=9...</td>
       </tr>
       <tr>
         <th>1</th>
         <td>P63313</td>
         <td>2</td>
         <td>0</td>
         <td>24.1939</td>
         <td>0.000000e+00</td>
         <td>0.000000</td>
         <td>3.823799</td>
         <td>1.000000</td>
         <td>COVID</td>
         <td>CTRL</td>
         <td>Thymosin beta-10 OS=Homo sapiens OX=9606 GN=TM...</td>
       </tr>
       <tr>
         <th>2</th>
         <td>P03886</td>
         <td>3</td>
         <td>0</td>
         <td>24.0213</td>
         <td>1.299387e-07</td>
         <td>0.000041</td>
         <td>1.386199</td>
         <td>0.999998</td>
         <td>CTRL</td>
         <td>COVID</td>
         <td>NADH-ubiquinone oxidoreductase chain 1 OS=Homo...</td>
       </tr>
       <tr>
         <th>3</th>
         <td>Q9BSM1</td>
         <td>2</td>
         <td>2</td>
         <td>12.2670</td>
         <td>5.516988e-07</td>
         <td>0.000105</td>
         <td>1.726615</td>
         <td>0.999984</td>
         <td>COVID</td>
         <td>CTRL</td>
         <td>Polycomb group RING finger protein 1 OS=Homo s...</td>
       </tr>
       <tr>
         <th>4</th>
         <td>O94819</td>
         <td>32</td>
         <td>16</td>
         <td>190.5708</td>
         <td>5.575815e-07</td>
         <td>0.000105</td>
         <td>1.245223</td>
         <td>0.999984</td>
         <td>COVID</td>
         <td>CTRL</td>
         <td>Kelch repeat and BTB domain-containing protein...</td>
       </tr>
       <tr>
         <th>...</th>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
         <td>...</td>
       </tr>
       <tr>
         <th>2273</th>
         <td>P20020</td>
         <td>67</td>
         <td>21</td>
         <td>523.2192</td>
         <td>9.977449e-01</td>
         <td>0.411967</td>
         <td>1.006182</td>
         <td>0.050001</td>
         <td>COVID</td>
         <td>CTRL</td>
         <td>Plasma membrane calcium-transporting ATPase 1 ...</td>
       </tr>
       <tr>
         <th>2274</th>
         <td>Q14240</td>
         <td>30</td>
         <td>8</td>
         <td>228.9400</td>
         <td>9.979637e-01</td>
         <td>0.411967</td>
         <td>1.115801</td>
         <td>0.050001</td>
         <td>COVID</td>
         <td>CTRL</td>
         <td>Eukaryotic initiation factor 4A-II OS=Homo sap...</td>
       </tr>
       <tr>
         <th>2275</th>
         <td>P11279</td>
         <td>8</td>
         <td>4</td>
         <td>49.3820</td>
         <td>9.980765e-01</td>
         <td>0.411967</td>
         <td>1.006122</td>
         <td>0.050001</td>
         <td>COVID</td>
         <td>CTRL</td>
         <td>Lysosome-associated membrane glycoprotein 1 OS...</td>
       </tr>
       <tr>
         <th>2276</th>
         <td>P02100</td>
         <td>5</td>
         <td>2</td>
         <td>53.0892</td>
         <td>9.997813e-01</td>
         <td>0.412321</td>
         <td>1.163175</td>
         <td>0.050000</td>
         <td>COVID</td>
         <td>CTRL</td>
         <td>Hemoglobin subunit epsilon OS=Homo sapiens OX=...</td>
       </tr>
       <tr>
         <th>2277</th>
         <td>O00194</td>
         <td>5</td>
         <td>1</td>
         <td>39.7453</td>
         <td>9.998110e-01</td>
         <td>0.412321</td>
         <td>1.052006</td>
         <td>0.050000</td>
         <td>COVID</td>
         <td>CTRL</td>
         <td>Ras-related protein Rab-27B OS=Homo sapiens OX...</td>
       </tr>
     </tbody>
   </table>
   <p>2278 rows × 11 columns</p>
   </div>


pdata
^^^^^

Pdata presents a description of each sample analysed. Pdata must have at least 3 columns: 'Sample', 'Condition', and 'Biological'.


#. **Sample:** Identifier for each sample analysed.
#. **Condition:** Respective group for each sample.
#. **Biological:** Respective biological replicates for each sample.

When performing longitudinal analysis, users must input a 'TimeCourse' column showing day/hour/time associated with the respective sample.

In order to provide a clearer understanding of how to construct a pdata, we have dedicated an entire section to its detailed description.

.. code-block:: python

   pdata = pd.read_excel('../tests/data/proteins/general.xlsx', sheet_name=2)
   pdata


.. raw:: html

   <div>
   <style scoped>
       .dataframe tbody tr th:only-of-type {
           vertical-align: middle;
       }

       .dataframe tbody tr th {
           vertical-align: top;
       }

       .dataframe thead th {
           text-align: right;
       }
   </style>
   <table border="1" class="dataframe">
     <thead>
       <tr style="text-align: right;">
         <th></th>
         <th>Sample</th>
         <th>Condition</th>
         <th>Biological</th>
       </tr>
     </thead>
     <tbody>
       <tr>
         <th>0</th>
         <td>VCC_HB_1_1_2020</td>
         <td>COVID</td>
         <td>1</td>
       </tr>
       <tr>
         <th>1</th>
         <td>VCC_HB_1_2</td>
         <td>COVID</td>
         <td>1</td>
       </tr>
       <tr>
         <th>2</th>
         <td>VCC_HB_2_1</td>
         <td>COVID</td>
         <td>2</td>
       </tr>
       <tr>
         <th>3</th>
         <td>VCC_HB_2_1_2</td>
         <td>COVID</td>
         <td>2</td>
       </tr>
       <tr>
         <th>4</th>
         <td>VCC_HB_3_1</td>
         <td>COVID</td>
         <td>3</td>
       </tr>
       <tr>
         <th>5</th>
         <td>VCC_HB_3_1_2</td>
         <td>COVID</td>
         <td>3</td>
       </tr>
       <tr>
         <th>6</th>
         <td>VCC_HB_4_1</td>
         <td>COVID</td>
         <td>4</td>
       </tr>
       <tr>
         <th>7</th>
         <td>VCC_HB_4_1_2</td>
         <td>COVID</td>
         <td>4</td>
       </tr>
       <tr>
         <th>8</th>
         <td>VCC_HB_5_1</td>
         <td>COVID</td>
         <td>5</td>
       </tr>
       <tr>
         <th>9</th>
         <td>VCC_HB_5_1_2</td>
         <td>COVID</td>
         <td>5</td>
       </tr>
       <tr>
         <th>10</th>
         <td>VCC_HB_6_1</td>
         <td>COVID</td>
         <td>6</td>
       </tr>
       <tr>
         <th>11</th>
         <td>VCC_HB_6_1_2</td>
         <td>COVID</td>
         <td>6</td>
       </tr>
       <tr>
         <th>12</th>
         <td>VCC_HB_7_1</td>
         <td>COVID</td>
         <td>7</td>
       </tr>
       <tr>
         <th>13</th>
         <td>VCC_HB_7_1_2</td>
         <td>COVID</td>
         <td>7</td>
       </tr>
       <tr>
         <th>14</th>
         <td>VCC_HB_8_1</td>
         <td>COVID</td>
         <td>8</td>
       </tr>
       <tr>
         <th>15</th>
         <td>VCC_HB_8_1_2</td>
         <td>COVID</td>
         <td>8</td>
       </tr>
       <tr>
         <th>16</th>
         <td>VCC_HB_9_1</td>
         <td>COVID</td>
         <td>9</td>
       </tr>
       <tr>
         <th>17</th>
         <td>VCC_HB_9_1_2</td>
         <td>COVID</td>
         <td>9</td>
       </tr>
       <tr>
         <th>18</th>
         <td>VCC_HB_10_1</td>
         <td>COVID</td>
         <td>10</td>
       </tr>
       <tr>
         <th>19</th>
         <td>VCC_HB_10_1_2_</td>
         <td>COVID</td>
         <td>10</td>
       </tr>
       <tr>
         <th>20</th>
         <td>VCC_HB_11_1</td>
         <td>COVID</td>
         <td>11</td>
       </tr>
       <tr>
         <th>21</th>
         <td>VCC_HB_11_1_2_</td>
         <td>COVID</td>
         <td>11</td>
       </tr>
       <tr>
         <th>22</th>
         <td>VCC_HB_12_1</td>
         <td>COVID</td>
         <td>12</td>
       </tr>
       <tr>
         <th>23</th>
         <td>VCC_HB_12_1_2_</td>
         <td>COVID</td>
         <td>12</td>
       </tr>
       <tr>
         <th>24</th>
         <td>VCC_HB_A_1</td>
         <td>CTRL</td>
         <td>1</td>
       </tr>
       <tr>
         <th>25</th>
         <td>VCC_HB_A_1_2</td>
         <td>CTRL</td>
         <td>1</td>
       </tr>
       <tr>
         <th>26</th>
         <td>VCC_HB_B_1</td>
         <td>CTRL</td>
         <td>2</td>
       </tr>
       <tr>
         <th>27</th>
         <td>VCC_HB_B_1_2</td>
         <td>CTRL</td>
         <td>2</td>
       </tr>
       <tr>
         <th>28</th>
         <td>VCC_HB_C_1</td>
         <td>CTRL</td>
         <td>3</td>
       </tr>
       <tr>
         <th>29</th>
         <td>VCC_HB_C_1_2</td>
         <td>CTRL</td>
         <td>3</td>
       </tr>
       <tr>
         <th>30</th>
         <td>VCC_HB_D_1</td>
         <td>CTRL</td>
         <td>4</td>
       </tr>
       <tr>
         <th>31</th>
         <td>VCC_HB_D_1_2</td>
         <td>CTRL</td>
         <td>4</td>
       </tr>
       <tr>
         <th>32</th>
         <td>VCC_HB_E_1</td>
         <td>CTRL</td>
         <td>5</td>
       </tr>
       <tr>
         <th>33</th>
         <td>VCC_HB_E_1_2</td>
         <td>CTRL</td>
         <td>5</td>
       </tr>
       <tr>
         <th>34</th>
         <td>VCC_HB_F_1</td>
         <td>CTRL</td>
         <td>6</td>
       </tr>
       <tr>
         <th>35</th>
         <td>VCC_HB_F_1_2</td>
         <td>CTRL</td>
         <td>6</td>
       </tr>
       <tr>
         <th>36</th>
         <td>VCC_HB_G_1</td>
         <td>CTRL</td>
         <td>7</td>
       </tr>
       <tr>
         <th>37</th>
         <td>VCC_HB_G_1_2</td>
         <td>CTRL</td>
         <td>7</td>
       </tr>
     </tbody>
   </table>
   </div>


Snapshot
--------

The Snapshot method is an alternative option in OmicScope that allows for the analysis of multiple studies (omics) by importing pre-analyzed data from other platforms.

To use the Snapshot method, the user needs to upload a CSV or Excel file organized as follows:


#. First row: **ControlGroup: INSERT_HERE_YOUR_CONTROL**
#. Second row: **Experimental: INSERT_HERE_YOUR_EXPERIMENTAL_GROUPS_SEPARATED_BY_COMMAS**
#. Subsequent rows: A table containing the following columns: 'Accession', 'gene_name', 'log2(fc)', and either 'pvalue' or 'pAdjusted'. This table structure is mandatory and must be included in file starting from the third row.

It is important to note that Snapshot contains a limited amount of information, which means that not all plots and enrichment analyses are available. However, once the data is imported into OmicScope, it can be exported as an .omics file and used in the Nebula module.

Additional Informations
-----------------------

Users can also define and optimize any extra parameters that are in the OmicScope function.


#. 
   **ControlGroup** (default = None): Users can define a control group ('ControlGroup=None', default) to perform comparisons against a specific group (this group has to be explicitly defined in the 'Conditions' column on the pdata table).

#. 
   **ExperimentalDesign** (default = 'static'): Comparisons among independent groups are called 'static' experimental designs. On the other hand, if the experiment takes into account several time points, then it is performing a 'longitudinal' experimental design (in this case, a pdata table must present a 'TimeCourse' column).

#. 
   **pvalue** (default = 'pAdjusted'): Defines the type of statistics used to report differentially regulated proteins. The options are nominal p-value ('pvalue'), Benjamini-Hochberg adjusted p-value ('pAdjusted'), or Tukey post-hoc correction ('pTukey', only for multiple group comparisons in static experiments).

#. 
   **PValue_cutoff** (default = 0.05): Statistical cutoff to consider proteins differentially regulated.

#. 
   **FoldChange_cutoff** (default = 0): Cutoff of abundance ratio to consider proteins differentially regulated.

#. 
   **logTransformed** (default = False): Usually, software reports abundance in nominal values, requiring a log-transformation of the values. If users perform transformation before the OmicScope workflow, set logTransformed=True.

#. 
   **ExcludeContaminants** (default = True): Since keratins are considered sample contaminants in most studies, OmicScope can exclude them from final results.

#. 
   **degrees_of_freedom** (default = 2): For longitudinal analysis, users can optimize the parameters according to their study, choosing a greater degree of freedom to perform the analysis.

#. 
   **independent_ttest** (default = True): If running a t-test, the user can specify if data sampling is independent (default) or paired (independent_ttest=False). Defaults to True.
