<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>2.6. Loading GFF files &mdash; machado 0.1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="2.7. Loading Feature Additional Info" href="load_feature_annotation.html" />
    <link rel="prev" title="2.5. Loading FASTA files" href="load_fasta.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> machado
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">1. Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="dataload.html">2. Data loading</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="load_ontologies.html">2.1. Ontologies</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_taxonomy.html">2.2. Taxonomy</a></li>
<li class="toctree-l2"><a class="reference internal" href="insert_organism.html">2.3. Organism</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_publication.html">2.4. Publication</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_fasta.html">2.5. FASTA</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">2.6. GFF</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_feature_annotation.html">2.7. Feature additional info</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_blast.html">2.8. Blast</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_interproscan.html">2.9. InterproScan</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_orthomcl.html">2.10. OrthoMCL</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_rnaseq.html">2.11. RNA-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_coexpression.html">2.12. Coexpression</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="visualization.html">3. Visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="diagrams.html">4. Diagrams</a></li>
<li class="toctree-l1"><a class="reference internal" href="models.html">5. Models</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">machado</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a></li>
          <li class="breadcrumb-item"><a href="dataload.html"><span class="section-number">2. </span>Data loading</a></li>
      <li class="breadcrumb-item active"><span class="section-number">2.6. </span>Loading GFF files</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/load_gff.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="loading-gff-files">
<h1><span class="section-number">2.6. </span>Loading GFF files<a class="headerlink" href="#loading-gff-files" title="Permalink to this heading"></a></h1>
<p>The first column of a GFF file is the reference sequence ID.
Usually, in order to load a GFF file, it’s required to have a reference FASTA file loaded.
But some GFF files already have the reference features such as chromosome or scaffold.
In this case, there are two options:</p>
<ul class="simple">
<li><p>Load the GFF directly, without a reference FASTA file</p></li>
<li><p>Load the FASTA file and then load the GFF using the parameter ‘ignore’ to not load the reference features</p></li>
</ul>
<p>The GFF file must be indexed using <a class="reference external" href="http://www.htslib.org/doc/tabix.html">tabix</a>.</p>
<section id="load-gff">
<h2><span class="section-number">2.6.1. </span>Load GFF<a class="headerlink" href="#load-gff" title="Permalink to this heading"></a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>manage.py<span class="w"> </span>load_gff<span class="w"> </span>--file<span class="w"> </span>organism_genes_sorted.gff3.gz<span class="w"> </span>--organism<span class="w"> </span><span class="s1">&#39;Arabidopsis thaliana&#39;</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Loading this file can be faster if you increase the number of threads (–cpu).</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>manage.py<span class="w"> </span>load_gff<span class="w"> </span>--help
</pre></div>
</div>
<table class="docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>–file</p></td>
<td><p>GFF3 genome file indexed with tabix (see <a class="reference external" href="http://www.htslib.org/doc/tabix.html">http://www.htslib.org/doc/tabix.html</a>) *</p></td>
</tr>
<tr class="row-even"><td><p>–organism</p></td>
<td><p>Species name (eg. Homo sapiens, Mus musculus) *</p></td>
</tr>
<tr class="row-odd"><td><p>–ignore</p></td>
<td><p>List of feature types to ignore (eg. chromosome scaffold)</p></td>
</tr>
<tr class="row-even"><td><p>–doi</p></td>
<td><p>DOI of a reference stored using <em>load_publication</em> (eg. 10.1111/s12122-012-1313-4)</p></td>
</tr>
<tr class="row-odd"><td><p>–qtl</p></td>
<td><p>Set this flag to handle GFF files from QTLDB</p></td>
</tr>
<tr class="row-even"><td><p>–cpu</p></td>
<td><p>Number of threads</p></td>
</tr>
</tbody>
</table>
<p>* required fields</p>
</section>
<section id="remove-file">
<h2><span class="section-number">2.6.2. </span>Remove file<a class="headerlink" href="#remove-file" title="Permalink to this heading"></a></h2>
<p>If, by any reason, you need to remove a GFF dataset you should use the command <em>remove_file</em>. <strong>If you delete a file, every record that depend on it will be deleted on cascade</strong>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>manage.py<span class="w"> </span>remove_file<span class="w"> </span>--help
</pre></div>
</div>
<ul class="simple">
<li><p>This command requires the file name (Dbxrefprop.value)</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="load_fasta.html" class="btn btn-neutral float-left" title="2.5. Loading FASTA files" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="load_feature_annotation.html" class="btn btn-neutral float-right" title="2.7. Loading Feature Additional Info" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2018, Embrapa.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>