<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>2.8. Loading Blast results &mdash; machado 0.1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="2.9. Loading InterproScan results" href="load_interproscan.html" />
    <link rel="prev" title="2.7. Loading Feature Additional Info" href="load_feature_annotation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> machado
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">1. Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="dataload.html">2. Data loading</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="load_ontologies.html">2.1. Ontologies</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_taxonomy.html">2.2. Taxonomy</a></li>
<li class="toctree-l2"><a class="reference internal" href="insert_organism.html">2.3. Organism</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_publication.html">2.4. Publication</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_fasta.html">2.5. FASTA</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_gff.html">2.6. GFF</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_feature_annotation.html">2.7. Feature additional info</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">2.8. Blast</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_interproscan.html">2.9. InterproScan</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_orthomcl.html">2.10. OrthoMCL</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_rnaseq.html">2.11. RNA-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_coexpression.html">2.12. Coexpression</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="visualization.html">3. Visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="diagrams.html">4. Diagrams</a></li>
<li class="toctree-l1"><a class="reference internal" href="models.html">5. Models</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">machado</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a></li>
          <li class="breadcrumb-item"><a href="dataload.html"><span class="section-number">2. </span>Data loading</a></li>
      <li class="breadcrumb-item active"><span class="section-number">2.8. </span>Loading Blast results</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/load_blast.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="loading-blast-results">
<h1><span class="section-number">2.8. </span>Loading Blast results<a class="headerlink" href="#loading-blast-results" title="Permalink to this heading"></a></h1>
<p>In order to load a BLAST xml result file, both the query and the subject records must be previously stored (see below). The current version was tested for loading BLAST analysis on <strong>proteins</strong>.</p>
<section id="load-blast-subject-records">
<h2><span class="section-number">2.8.1. </span>Load BLAST subject records<a class="headerlink" href="#load-blast-subject-records" title="Permalink to this heading"></a></h2>
<p>In case you did a BLAST against a multispecies protein database, like NCBI’s nr or Uniprot’s trembl or swissprot, you need to load previously all subject matches before loading the result itself. To do so, use the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>manage.py<span class="w"> </span>load_similarity_matches<span class="w"> </span>--file<span class="w"> </span>blast_result.xml<span class="w"> </span>--format<span class="w"> </span>blast-xml
</pre></div>
</div>
<ul class="simple">
<li><p>Loading this file can be faster if you increase the number of threads (–cpu).</p></li>
</ul>
</section>
<section id="load-blast">
<h2><span class="section-number">2.8.2. </span>Load BLAST<a class="headerlink" href="#load-blast" title="Permalink to this heading"></a></h2>
<p>If all queries and subjects are already loaded you can run the following command to load a BLAST xml result:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>manage.py<span class="w"> </span>load_similarity<span class="w"> </span>--file<span class="w"> </span>blast_result.xml<span class="w"> </span>--format<span class="w"> </span>blast-xml<span class="w"> </span>--so_query<span class="w"> </span>polypeptide<span class="w"> </span>--so_subject<span class="w"> </span>protein_match<span class="w"> </span>--program<span class="w"> </span>diamond<span class="w"> </span>--programversion<span class="w"> </span><span class="m">0</span>.9.24<span class="w"> </span>--organism_query<span class="w"> </span><span class="s1">&#39;Oryza sativa&#39;</span><span class="w"> </span>--organism_subject<span class="w"> </span><span class="s1">&#39;multispecies multispecies&#39;</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Loading this file can be faster if you increase the number of threads (–cpu).</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>manage.py<span class="w"> </span>load_similarity<span class="w"> </span>--help
</pre></div>
</div>
<table class="docutils align-default">
<colgroup>
<col style="width: 14%" />
<col style="width: 86%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>–file</p></td>
<td><p>Blast XML file *</p></td>
</tr>
<tr class="row-even"><td><p>–format</p></td>
<td><p>blast-xml *</p></td>
</tr>
<tr class="row-odd"><td><p>–so_query</p></td>
<td><p>Query Sequence Ontology term. (eg. assembly, mRNA, polypeptide) *</p></td>
</tr>
<tr class="row-even"><td><p>–so_subject</p></td>
<td><p>Subject Sequence Ontology term. (eg. protein_match if loading a BLAST result) *</p></td>
</tr>
<tr class="row-odd"><td><p>–organism_query</p></td>
<td><p>Query’s organism name. eg. ‘Oryza sativa’. Cannot be ‘multispecies’ *</p></td>
</tr>
<tr class="row-even"><td><p>–organism_subject</p></td>
<td><p>Subject’s organism name eg. ‘Oryza sativa’. Put ‘multispecies multispecies’ if using a multispecies database. *</p></td>
</tr>
<tr class="row-odd"><td><p>–program</p></td>
<td><p>Program *</p></td>
</tr>
<tr class="row-even"><td><p>–programversion</p></td>
<td><p>Program version *</p></td>
</tr>
<tr class="row-odd"><td><p>–name</p></td>
<td><p>Name</p></td>
</tr>
<tr class="row-even"><td><p>–description</p></td>
<td><p>Description</p></td>
</tr>
<tr class="row-odd"><td><p>–algorithm</p></td>
<td><p>Algorithm</p></td>
</tr>
<tr class="row-even"><td><p>–cpu</p></td>
<td><p>Number of threads</p></td>
</tr>
</tbody>
</table>
<p>* required fields</p>
</section>
<section id="remove-file">
<h2><span class="section-number">2.8.3. </span>Remove file<a class="headerlink" href="#remove-file" title="Permalink to this heading"></a></h2>
<p>If, by any reason, you need to remove a Blast result set you should use the command <em>remove_analysis</em>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>manage.py<span class="w"> </span>remove_analysis<span class="w"> </span>--help
</pre></div>
</div>
<ul class="simple">
<li><p>This command requires the analysis name (Analysis.sourcename)</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="load_feature_annotation.html" class="btn btn-neutral float-left" title="2.7. Loading Feature Additional Info" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="load_interproscan.html" class="btn btn-neutral float-right" title="2.9. Loading InterproScan results" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2018, Embrapa.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>