

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Index &mdash; TRANSIT v1.0.5 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/default.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="#" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> TRANSIT
          

          
            
            <img src="_static/transit_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                v1.0.5
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">TRANSIT2 MANUAL</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="transit_overview.html">TRANSIT2 Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="transit_overview.html#typical-workflow">Typical Workflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit_overview.html#tpp-transit-pre-processor">TPP (Transit Pre-Processor)</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit_overview.html#combined-wig-and-metadata-files">Combined_wig and Metadata Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit_overview.html#genome-annotations-prot-tables-and-gff-files">Genome Annotations (.prot_tables and .gff files)</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit_overview.html#gui">GUI</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit_overview.html#pre-processing-tasks">Pre-Processing Tasks</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit_overview.html#statistical-analysis-methods">Statistical Analysis Methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit_overview.html#analyses-for-tn5-datasets">Analyses for Tn5 Datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit_overview.html#results-and-post-processing">Results and Post-Processing</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit_overview.html#command-line">Command Line</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit_overview.html#developers">Developers</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit_overview.html#references">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="transit_install.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="transit_install.html#installing-transit2-using-pip">Installing Transit2 using Pip</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit_install.html#installing-transit2-using-git">Installing Transit2 using Git</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit_install.html#requirements">Requirements</a><ul>
<li class="toctree-l3"><a class="reference internal" href="transit_install.html#additional-requirements-r-statistical-analysis-package">Additional Requirements: R (statistical analysis package)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="transit_install.html#optional-install-bwa-to-use-with-tpp-pre-processor">Optional: Install BWA to use with TPP pre-processor</a><ul>
<li class="toctree-l3"><a class="reference internal" href="transit_install.html#linux-osx-instructions">Linux &amp; OSX Instructions</a></li>
<li class="toctree-l3"><a class="reference internal" href="transit_install.html#windows-instructions">Windows Instructions</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="transit_install.html#upgrading">Upgrading</a><ul>
<li class="toctree-l3"><a class="reference internal" href="transit_install.html#method-1-upgrading-package-installation">Method 1: Upgrading package installation</a></li>
<li class="toctree-l3"><a class="reference internal" href="transit_install.html#method-2-upgrading-source-installation">Method 2: Upgrading source installation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="transit_install.html#installing-wxpython">Installing wxPython</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit_install.html#troubleshooting">Troubleshooting</a><ul>
<li class="toctree-l3"><a class="reference internal" href="transit_install.html#no-window-appears-when-running-in-gui-mode">1. No window appears when running in GUI mode.</a></li>
<li class="toctree-l3"><a class="reference internal" href="transit_install.html#pip3-systemerror-cannot-compile-python-h">2. pip3: SystemError: Cannot compile ‘Python.h’.</a></li>
<li class="toctree-l3"><a class="reference internal" href="transit_install.html#pip-the-following-required-packages-can-not-be-built-freetype-png-etc">3. pip: “The following required packages can not be built: freetype,png,” etc.</a></li>
<li class="toctree-l3"><a class="reference internal" href="transit_install.html#pip3-no-lapack-blas-resources-found">4. pip3: “No lapack/blas resources found”</a></li>
<li class="toctree-l3"><a class="reference internal" href="transit_install.html#resources-contextualversionconflict-six-1-5-2">5. “resources.ContextualVersionConflict (six 1.5.2)…”</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="transit_running.html">Running TRANSIT</a><ul>
<li class="toctree-l2"><a class="reference internal" href="transit_running.html#gui-mode">GUI Mode</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit_running.html#command-line-mode">Command line Mode</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="input_files.html">Input Files for Transit</a><ul>
<li class="toctree-l2"><a class="reference internal" href="input_files.html#combined-wig-files">Combined_wig Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="input_files.html#samples-metadata-file">Samples Metadata File</a></li>
<li class="toctree-l2"><a class="reference internal" href="input_files.html#genome-annotations-prot-tables-and-gff-files">Genome Annotations (.prot_tables and .gff files)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="quality_control.html">Quality Control</a><ul>
<li class="toctree-l2"><a class="reference internal" href="quality_control.html#qc-metrics-table">QC Metrics Table</a></li>
<li class="toctree-l2"><a class="reference internal" href="quality_control.html#qc-plots">QC Plots</a><ul>
<li class="toctree-l3"><a class="reference internal" href="quality_control.html#figure-1-read-count-distribution">Figure 1: Read-Count Distribution</a></li>
<li class="toctree-l3"><a class="reference internal" href="quality_control.html#figure-2-qq-plot-of-read-counts-vs-geometric-distribution">Figure 2: QQ-Plot of Read-Counts vs Geometric Distribution</a></li>
<li class="toctree-l3"><a class="reference internal" href="quality_control.html#figure-3-ranked-plot-of-read-counts">Figure 3: Ranked plot of Read-Counts</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="quality_control.html#interpretation-of-data-quality">Interpretation of Data Quality</a></li>
<li class="toctree-l2"><a class="reference internal" href="quality_control.html#what-to-do-about-skewed-samples">What to do about Skewed Samples?</a></li>
<li class="toctree-l2"><a class="reference internal" href="quality_control.html#beta-geometric-correction">Beta-Geometric Correction</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">PRE-PROCESSING</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="tnseq_stats_method.html">Tnseq_stats</a></li>
<li class="toctree-l1"><a class="reference internal" href="scatterplot_method.html">Scatterplot</a><ul>
<li class="toctree-l2"><a class="reference internal" href="scatterplot_method.html#command-line-usage">Command Line Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="scatterplot_method.html#example-usage">Example Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="scatterplot_method.html#gui-mode">GUI Mode</a></li>
<li class="toctree-l2"><a class="reference internal" href="scatterplot_method.html#runtime">Runtime</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="corrplot_method.html">Corrplot</a><ul>
<li class="toctree-l2"><a class="reference internal" href="corrplot_method.html#command-line-usage">Command Line Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="corrplot_method.html#gui-mode">GUI Mode</a></li>
<li class="toctree-l2"><a class="reference internal" href="corrplot_method.html#runtime">Runtime</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="normalize_method.html">Normalization</a><ul>
<li class="toctree-l2"><a class="reference internal" href="normalize_method.html#command-line">Command-line</a></li>
<li class="toctree-l2"><a class="reference internal" href="normalize_method.html#gui">GUI</a></li>
<li class="toctree-l2"><a class="reference internal" href="normalize_method.html#examples">Examples</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="gene_means_method.html">Gene Means</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">ANALYSES of single conditions</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="gumbel_method.html">Gumbel</a><ul>
<li class="toctree-l2"><a class="reference internal" href="gumbel_method.html#how-does-it-work">How does it work?</a></li>
<li class="toctree-l2"><a class="reference internal" href="gumbel_method.html#usage">Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="gumbel_method.html#parameters">Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="gumbel_method.html#gui-mode">GUI Mode</a></li>
<li class="toctree-l2"><a class="reference internal" href="gumbel_method.html#outputs-and-diagnostics">Outputs and diagnostics</a></li>
<li class="toctree-l2"><a class="reference internal" href="gumbel_method.html#run-time">Run-time</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="hmm_method.html">HMM</a><ul>
<li class="toctree-l2"><a class="reference internal" href="hmm_method.html#how-does-it-work">How does it work?</a></li>
<li class="toctree-l2"><a class="reference internal" href="hmm_method.html#usage">Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="hmm_method.html#parameters">Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="hmm_method.html#gui-mode">GUI Mode</a></li>
<li class="toctree-l2"><a class="reference internal" href="hmm_method.html#output-and-diagnostics">Output and Diagnostics</a><ul>
<li class="toctree-l3"><a class="reference internal" href="hmm_method.html#sites-output-file">Sites Output File:</a></li>
<li class="toctree-l3"><a class="reference internal" href="hmm_method.html#genes-output-file">Genes Output File:</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="hmm_method.html#run-time">Run-time</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="ttnfitness_method.html">TTN-Fitness</a><ul>
<li class="toctree-l2"><a class="reference internal" href="ttnfitness_method.html#usage">Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="ttnfitness_method.html#gui-mode">GUI Mode</a></li>
<li class="toctree-l2"><a class="reference internal" href="ttnfitness_method.html#volcano-plot">Volcano Plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="ttnfitness_method.html#output">Output</a><ul>
<li class="toctree-l3"><a class="reference internal" href="ttnfitness_method.html#genes-output-file">Genes Output File:</a></li>
<li class="toctree-l3"><a class="reference internal" href="ttnfitness_method.html#sites-output-file">Sites Output File:</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="ttnfitness_method.html#example-of-running-the-ttn-fitness-methodology-on-the-sample-glycerol-data">Example of running the TTN-Fitness methodology on the sample glycerol data</a></li>
<li class="toctree-l2"><a class="reference internal" href="ttnfitness_method.html#run-time">Run-time</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">ANALYSES for pairwise comparisons</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="resampling_method.html">Resampling</a><ul>
<li class="toctree-l2"><a class="reference internal" href="resampling_method.html#how-does-it-work">How does it work?</a></li>
<li class="toctree-l2"><a class="reference internal" href="resampling_method.html#site-restricted-s-r-resampling">Site-Restricted (S-R) Resampling</a></li>
<li class="toctree-l2"><a class="reference internal" href="resampling_method.html#usage">Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="resampling_method.html#parameters">Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="resampling_method.html#notes">Notes</a></li>
<li class="toctree-l2"><a class="reference internal" href="resampling_method.html#command-line-examples">Command Line Examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="resampling_method.html#doing-resampling-with-datasets-from-different-libraries">Doing resampling with datasets from different libraries</a></li>
<li class="toctree-l2"><a class="reference internal" href="resampling_method.html#doing-resampling-between-different-strains">Doing resampling between different strains</a></li>
<li class="toctree-l2"><a class="reference internal" href="resampling_method.html#gui-mode">GUI Mode</a></li>
<li class="toctree-l2"><a class="reference internal" href="resampling_method.html#output-and-diagnostics">Output and Diagnostics</a></li>
<li class="toctree-l2"><a class="reference internal" href="resampling_method.html#run-time">Run-time</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="utest_method.html">Mann-Whitney U-test (utest)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="utest_method.html#usage">Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="utest_method.html#gui-mode">GUI Mode</a></li>
<li class="toctree-l2"><a class="reference internal" href="utest_method.html#output">Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="utest_method.html#runtime">Runtime</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">ANALYSES for multiple conditions</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="gi_method.html">Genetic Interactions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="gi_method.html#how-does-it-work">How does it work?</a></li>
<li class="toctree-l2"><a class="reference internal" href="gi_method.html#statistical-significance">Statistical Significance</a></li>
<li class="toctree-l2"><a class="reference internal" href="gi_method.html#usage">Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="gi_method.html#example">Example</a></li>
<li class="toctree-l2"><a class="reference internal" href="gi_method.html#parameters">Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="gi_method.html#gui-mode">Gui Mode</a></li>
<li class="toctree-l2"><a class="reference internal" href="gi_method.html#output-and-diagnostics">Output and Diagnostics</a></li>
<li class="toctree-l2"><a class="reference internal" href="gi_method.html#run-time">Run-time</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="anova_method.html">ANOVA</a><ul>
<li class="toctree-l2"><a class="reference internal" href="anova_method.html#how-does-it-work">How does it work?</a></li>
<li class="toctree-l2"><a class="reference internal" href="anova_method.html#example">Example</a></li>
<li class="toctree-l2"><a class="reference internal" href="anova_method.html#input-files">Input files</a></li>
<li class="toctree-l2"><a class="reference internal" href="anova_method.html#parameters">Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="anova_method.html#output-and-diagnostics">Output and Diagnostics</a></li>
<li class="toctree-l2"><a class="reference internal" href="anova_method.html#gui-mode">GUI Mode</a></li>
<li class="toctree-l2"><a class="reference internal" href="anova_method.html#run-time">Run-time</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="zinb_method.html">ZINB</a><ul>
<li class="toctree-l2"><a class="reference internal" href="zinb_method.html#how-does-it-work">How does it work?</a></li>
<li class="toctree-l2"><a class="reference internal" href="zinb_method.html#example">Example</a></li>
<li class="toctree-l2"><a class="reference internal" href="zinb_method.html#input-files">Input files</a></li>
<li class="toctree-l2"><a class="reference internal" href="zinb_method.html#parameters">Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="zinb_method.html#covariates-and-interactions">Covariates and Interactions</a></li>
<li class="toctree-l2"><a class="reference internal" href="zinb_method.html#categorical-vs-numeric-covariates">Categorical vs Numeric Covariates</a></li>
<li class="toctree-l2"><a class="reference internal" href="zinb_method.html#statistical-significance-what-the-p-values-mean-in-the-zinb-output">Statistical Significance - What the P-values Mean in the ZINB Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="zinb_method.html#gui-mode">GUI Mode</a></li>
<li class="toctree-l2"><a class="reference internal" href="zinb_method.html#output-and-diagnostics">Output and Diagnostics</a></li>
<li class="toctree-l2"><a class="reference internal" href="zinb_method.html#run-time">Run-time</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">POST-PROCESSING</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pathway_enrichment_method.html">Pathway Enrichment Analysis</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pathway_enrichment_method.html#command-line-usage">Command Line Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="pathway_enrichment_method.html#parameters">Parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="pathway_enrichment_method.html#auxilliary-pathway-files-in-transit-data-directory">Auxilliary Pathway Files in Transit Data Directory</a></li>
<li class="toctree-l2"><a class="reference internal" href="pathway_enrichment_method.html#current-recommendations">Current Recommendations</a></li>
<li class="toctree-l2"><a class="reference internal" href="pathway_enrichment_method.html#examples">Examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="pathway_enrichment_method.html#gui-mode">GUI Mode</a></li>
<li class="toctree-l2"><a class="reference internal" href="pathway_enrichment_method.html#output-and-diagnostics">Output and Diagnostics</a></li>
<li class="toctree-l2"><a class="reference internal" href="pathway_enrichment_method.html#run-time">Run-time</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="heatmap_method.html">Heatmap</a><ul>
<li class="toctree-l2"><a class="reference internal" href="heatmap_method.html#command-line-usage">Command Line Usage:</a></li>
<li class="toctree-l2"><a class="reference internal" href="heatmap_method.html#gui">GUI:</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">TPP Manual</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="tpp.html">TPP Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#running-tpp">Running TPP</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#mapping-to-genomes-with-multiple-contigs">Mapping to Genomes with Multiple Contigs</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#overview-of-data-processing-procedure">Overview of Data Processing Procedure</a></li>
<li class="toctree-l1"><a class="reference internal" href="tpp.html#statistics">Statistics</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Code Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="transit.html">transit package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="transit.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit.html#pytransit-specific-tools-norm-tools-module">pytransit.specific_tools.norm_tools module</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit.html#pytransit-specific-tools-stat-tools-module">pytransit.specific_tools.stat_tools module</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit.html#pytransit-specific-tools-tnseq-tools-module">pytransit.specific_tools.tnseq_tools module</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit.html#pytransit-specific-tools-transit-tools-module">pytransit.specific_tools.transit_tools module</a></li>
<li class="toctree-l2"><a class="reference internal" href="transit.html#module-pytransit">Module contents</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">TRANSIT</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Index</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            

<h1 id="index">Index</h1>

<div class="genindex-jumpbox">
 <a href="#_"><strong>_</strong></a>
 | <a href="#A"><strong>A</strong></a>
 | <a href="#B"><strong>B</strong></a>
 | <a href="#C"><strong>C</strong></a>
 | <a href="#D"><strong>D</strong></a>
 | <a href="#E"><strong>E</strong></a>
 | <a href="#F"><strong>F</strong></a>
 | <a href="#G"><strong>G</strong></a>
 | <a href="#H"><strong>H</strong></a>
 | <a href="#I"><strong>I</strong></a>
 | <a href="#L"><strong>L</strong></a>
 | <a href="#M"><strong>M</strong></a>
 | <a href="#N"><strong>N</strong></a>
 | <a href="#O"><strong>O</strong></a>
 | <a href="#P"><strong>P</strong></a>
 | <a href="#Q"><strong>Q</strong></a>
 | <a href="#R"><strong>R</strong></a>
 | <a href="#S"><strong>S</strong></a>
 | <a href="#T"><strong>T</strong></a>
 | <a href="#V"><strong>V</strong></a>
 | <a href="#W"><strong>W</strong></a>
 | <a href="#Z"><strong>Z</strong></a>
 
</div>
<h2 id="_">_</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.__contains__">__contains__() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.__eq__">__eq__() (pytransit.specific_tools.tnseq_tools.Gene method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.__getitem__">__getitem__() (pytransit.specific_tools.tnseq_tools.Gene method)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.__getitem__">(pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.__len__">__len__() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.__lt__">__lt__() (pytransit.specific_tools.tnseq_tools.Gene method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.__str__">__str__() (pytransit.specific_tools.tnseq_tools.Gene method)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.__str__">(pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      </ul></li>
  </ul></td>
</tr></table>

<h2 id="A">A</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.AdaptiveBGCNorm">AdaptiveBGCNorm (class in pytransit.specific_tools.norm_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.AnnotationFile">AnnotationFile (class in pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.AnnotationFile.as_list_of_dicts">as_list_of_dicts (pytransit.specific_tools.tnseq_tools.AnnotationFile property)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile.as_prot_table_rows">as_prot_table_rows() (pytransit.specific_tools.tnseq_tools.GffFile method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.AssumeZerosDialog">AssumeZerosDialog (class in pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile.attributes_index">attributes_index (pytransit.specific_tools.tnseq_tools.GffFile attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.averaged">averaged() (pytransit.specific_tools.tnseq_tools.CombinedWig method)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="B">B</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.b_fdr">b_fdr() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.basename">basename() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.bayesian_essentiality_thresholds">bayesian_essentiality_thresholds() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.bernoulli_diff_distribution">bernoulli_diff_distribution() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.BetaGeomNorm">BetaGeomNorm (class in pytransit.specific_tools.norm_tools)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.bh_fdr_correction">bh_fdr_correction() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.binom">binom() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.binom_cdf">binom_cdf() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.binom_test">binom_test() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.boxcox_table">boxcox_table() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.boxcox_transform">boxcox_transform() (in module pytransit.specific_tools.stat_tools)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="C">C</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.c_terminus">c_terminus (pytransit.specific_tools.tnseq_tools.Genes attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.calc_gene_means">calc_gene_means() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="pytransit.html#pytransit.transit_gui.PngFile.can_load">can_load() (pytransit.transit_gui.PngFile static method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.check_wig_includes_zeros">check_wig_includes_zeros() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.clean_from_geometric">clean_from_geometric() (in module pytransit.specific_tools.norm_tools)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.AdaptiveBGCNorm.clean_from_geometric">(pytransit.specific_tools.norm_tools.AdaptiveBGCNorm method)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.BetaGeomNorm.clean_from_geometric">(pytransit.specific_tools.norm_tools.BetaGeomNorm method)</a>
</li>
      </ul></li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.column">column() (pytransit.specific_tools.tnseq_tools.CombinedWigMetadata method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.comb">comb() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.combine_lib_dicts">combine_lib_dicts() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.combine_replicates">combine_replicates() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig">CombinedWig (class in pytransit.specific_tools.tnseq_tools)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigData">CombinedWigData (class in pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata">CombinedWigMetadata (class in pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Condition">Condition (class in pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.condition_names">condition_names (pytransit.specific_tools.tnseq_tools.CombinedWig property)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.condition_names">(pytransit.specific_tools.tnseq_tools.CombinedWigMetadata property)</a>
</li>
      </ul></li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.condition_names_for">condition_names_for() (pytransit.specific_tools.tnseq_tools.CombinedWigMetadata method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.conditions">conditions (pytransit.specific_tools.tnseq_tools.CombinedWig property)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.conditions_by_wig_fingerprint">conditions_by_wig_fingerprint (pytransit.specific_tools.tnseq_tools.CombinedWigMetadata property)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.copy">copy() (pytransit.specific_tools.tnseq_tools.CombinedWig method)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.copy">(pytransit.specific_tools.tnseq_tools.CombinedWigMetadata method)</a>
</li>
      </ul></li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.covariates_by_wig_fingerprint_list">covariates_by_wig_fingerprint_list (pytransit.specific_tools.tnseq_tools.CombinedWigMetadata property)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.covariates_dict_for">covariates_dict_for() (pytransit.specific_tools.tnseq_tools.CombinedWigMetadata method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.cumulative_average">cumulative_average() (in module pytransit.specific_tools.stat_tools)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="D">D</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.d_binomial_diff_distribution">d_binomial_diff_distribution() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.desc">desc (pytransit.specific_tools.tnseq_tools.Gene attribute)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.dirname">dirname() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.TimerAndOutputs.duration_in_seconds">duration_in_seconds (pytransit.specific_tools.transit_tools.TimerAndOutputs property)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="E">E</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.ecdf">ecdf() (in module pytransit.specific_tools.norm_tools)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.AdaptiveBGCNorm.ecdf">(pytransit.specific_tools.norm_tools.AdaptiveBGCNorm method)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.BetaGeomNorm.ecdf">(pytransit.specific_tools.norm_tools.BetaGeomNorm method)</a>
</li>
      </ul></li>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.EmpHistNorm">EmpHistNorm (class in pytransit.specific_tools.norm_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.empirical_theta">empirical_theta() (in module pytransit.specific_tools.norm_tools)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.TTRNorm.empirical_theta">(pytransit.specific_tools.norm_tools.TTRNorm method)</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.end">end (pytransit.specific_tools.tnseq_tools.Gene attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile.end_index">end_index (pytransit.specific_tools.tnseq_tools.GffFile attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.expand_var">expand_var() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.expected_runs">expected_runs() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile.extract_gene_info">extract_gene_info() (pytransit.specific_tools.tnseq_tools.GffFile static method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.extract_yaml_data">extract_yaml_data() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="F">F</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.f_mean_diff_dict">f_mean_diff_dict() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.f_mean_diff_flat">f_mean_diff_flat() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.f_shuffle_dict_libraries">f_shuffle_dict_libraries() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.f_shuffle_flat">f_shuffle_flat() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.f_sum_diff_dict">f_sum_diff_dict() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.f_sum_diff_flat">f_sum_diff_flat() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.fact">fact() (in module pytransit.specific_tools.stat_tools)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.fetch_name">fetch_name() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile.file_extensions">file_extensions (pytransit.specific_tools.tnseq_tools.GffFile attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.file_starts_with">file_starts_with() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.filter_wigs_by_conditions">filter_wigs_by_conditions() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.filter_wigs_by_conditions2">filter_wigs_by_conditions2() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.fingerprint_for">fingerprint_for() (pytransit.specific_tools.tnseq_tools.CombinedWigMetadata method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.fingerprints_for">fingerprints_for() (pytransit.specific_tools.tnseq_tools.CombinedWigMetadata method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.fwer_bayes">fwer_bayes() (in module pytransit.specific_tools.stat_tools)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="G">G</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.gather_sample_data_for">gather_sample_data_for() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.gather_wig_data">gather_wig_data() (pytransit.specific_tools.tnseq_tools.CombinedWig static method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene">Gene (class in pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.AnnotationFile.gene_description">gene_description() (pytransit.specific_tools.tnseq_tools.AnnotationFile method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes">Genes (class in pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.genes">genes (pytransit.specific_tools.tnseq_tools.Genes attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_coordinate_map">get_coordinate_map() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_data_stats">get_data_stats() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_data_w_genome">get_data_w_genome() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_data_zero_fill">get_data_zero_fill() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_e1">get_e1() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_e2">get_e2() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_file_types">get_file_types() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_gamma">get_gamma() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.get_gap_span">get_gap_span() (pytransit.specific_tools.tnseq_tools.Gene method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.get_gene_span">get_gene_span() (pytransit.specific_tools.tnseq_tools.Gene method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.get_genes">get_genes() (pytransit.specific_tools.tnseq_tools.CombinedWig method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_genes_in_range">get_genes_in_range() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.get_lib_data_dict">get_lib_data_dict() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_pos_hash">get_pos_hash() (in module pytransit.specific_tools.tnseq_tools)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.transit_tools.get_pos_hash">(in module pytransit.specific_tools.transit_tools)</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_pos_hash_gff">get_pos_hash_gff() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_pos_hash_pt">get_pos_hash_pt() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_r1">get_r1() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_r2">get_r2() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.get_samples_metadata">get_samples_metadata() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.get_stats_by_rv">get_stats_by_rv() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_unknown_file_types">get_unknown_file_types() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.get_validated_data">get_validated_data() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.get_wig_stats">get_wig_stats() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile">GffFile (class in pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile.GffRow">GffRow (pytransit.specific_tools.tnseq_tools.GffFile attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.global_insertion">global_insertion() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.global_phi">global_phi() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.global_reads">global_reads() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.global_run">global_run() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.global_sites">global_sites() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.global_theta">global_theta() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.griffin_analysis">griffin_analysis() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.gumbel_cdf">gumbel_cdf() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="H">H</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.hdi_from_mcmc">hdi_from_mcmc() (in module pytransit.specific_tools.stat_tools)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="I">I</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.id_for">id_for() (pytransit.specific_tools.tnseq_tools.CombinedWigMetadata method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.ignore_codon">ignore_codon (pytransit.specific_tools.tnseq_tools.Genes attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.include_nc">include_nc (pytransit.specific_tools.tnseq_tools.Genes attribute)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.AnnotationFile.Info">Info (pytransit.specific_tools.tnseq_tools.AnnotationFile attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.interactions_by_wig_fingerprint_list">interactions_by_wig_fingerprint_list (pytransit.specific_tools.tnseq_tools.CombinedWigMetadata property)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile.is_definitely_not_gff3">is_definitely_not_gff3() (pytransit.specific_tools.tnseq_tools.GffFile static method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.is_even">is_even() (in module pytransit.specific_tools.stat_tools)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="L">L</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.load">load() (pytransit.specific_tools.tnseq_tools.CombinedWig static method)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigData.load">(pytransit.specific_tools.tnseq_tools.CombinedWigData static method)</a>
</li>
      </ul></li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.local_gap_span">local_gap_span() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.local_gene_span">local_gene_span() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.local_insertions">local_insertions() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.local_phis">local_phis() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.local_reads">local_reads() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.local_runs">local_runs() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.local_sites">local_sites() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.local_thetas">local_thetas() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.loess">loess() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.loess_correction">loess_correction() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.log_fac">log_fac() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.log_likelihood">log_likelihood() (in module pytransit.specific_tools.stat_tools)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="M">M</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile.max_number_of_columns">max_number_of_columns (pytransit.specific_tools.tnseq_tools.GffFile attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.max_run">max_run() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.minread">minread (pytransit.specific_tools.tnseq_tools.Genes attribute)</a>
</li>
      <li>
    module

      <ul>
        <li><a href="pytransit.html#module-pytransit">pytransit</a>, <a href="transit.html#module-pytransit">[1]</a>
</li>
        <li><a href="transit.html#module-pytransit.specific_tools.norm_tools">pytransit.specific_tools.norm_tools</a>
</li>
        <li><a href="transit.html#module-pytransit.specific_tools.stat_tools">pytransit.specific_tools.stat_tools</a>
</li>
        <li><a href="transit.html#module-pytransit.specific_tools.tnseq_tools">pytransit.specific_tools.tnseq_tools</a>
</li>
        <li><a href="transit.html#module-pytransit.specific_tools.transit_tools">pytransit.specific_tools.transit_tools</a>
</li>
        <li><a href="pytransit.html#module-pytransit.transit_gui">pytransit.transit_gui</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.multinomial">multinomial() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.my_perm">my_perm() (in module pytransit.specific_tools.stat_tools)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="N">N</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.n_terminus">n_terminus (pytransit.specific_tools.tnseq_tools.Genes attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.AdaptiveBGCNorm.name">name (pytransit.specific_tools.norm_tools.AdaptiveBGCNorm attribute)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.BetaGeomNorm.name">(pytransit.specific_tools.norm_tools.BetaGeomNorm attribute)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.EmpHistNorm.name">(pytransit.specific_tools.norm_tools.EmpHistNorm attribute)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.NoNorm.name">(pytransit.specific_tools.norm_tools.NoNorm attribute)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.NormMethod.name">(pytransit.specific_tools.norm_tools.NormMethod attribute)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.NZMeanNorm.name">(pytransit.specific_tools.norm_tools.NZMeanNorm attribute)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.QuantileNorm.name">(pytransit.specific_tools.norm_tools.QuantileNorm attribute)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.TotReadsNorm.name">(pytransit.specific_tools.norm_tools.TotReadsNorm attribute)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.TTRNorm.name">(pytransit.specific_tools.norm_tools.TTRNorm attribute)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.ZeroInflatedNBNorm.name">(pytransit.specific_tools.norm_tools.ZeroInflatedNBNorm attribute)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.name">(pytransit.specific_tools.tnseq_tools.Gene attribute)</a>
</li>
      </ul></li>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.NoNorm">NoNorm (class in pytransit.specific_tools.norm_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.norm">norm (pytransit.specific_tools.tnseq_tools.Genes attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.norm">norm() (in module pytransit.specific_tools.stat_tools)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.norm_to_target">norm_to_target() (in module pytransit.specific_tools.norm_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.AdaptiveBGCNorm.normalize">normalize() (pytransit.specific_tools.norm_tools.AdaptiveBGCNorm static method)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.BetaGeomNorm.normalize">(pytransit.specific_tools.norm_tools.BetaGeomNorm static method)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.EmpHistNorm.normalize">(pytransit.specific_tools.norm_tools.EmpHistNorm static method)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.NoNorm.normalize">(pytransit.specific_tools.norm_tools.NoNorm static method)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.NormMethod.normalize">(pytransit.specific_tools.norm_tools.NormMethod static method)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.NZMeanNorm.normalize">(pytransit.specific_tools.norm_tools.NZMeanNorm static method)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.QuantileNorm.normalize">(pytransit.specific_tools.norm_tools.QuantileNorm static method)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.TotReadsNorm.normalize">(pytransit.specific_tools.norm_tools.TotReadsNorm static method)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.TTRNorm.normalize">(pytransit.specific_tools.norm_tools.TTRNorm static method)</a>
</li>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.ZeroInflatedNBNorm.normalize">(pytransit.specific_tools.norm_tools.ZeroInflatedNBNorm static method)</a>
</li>
      </ul></li>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.normalize_data">normalize_data() (in module pytransit.specific_tools.norm_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.normalized_with">normalized_with() (pytransit.specific_tools.tnseq_tools.CombinedWig method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.NormMethod">NormMethod (class in pytransit.specific_tools.norm_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.NZMeanNorm">NZMeanNorm (class in pytransit.specific_tools.norm_tools)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="O">O</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.AssumeZerosDialog.OnClose">OnClose() (pytransit.specific_tools.transit_tools.AssumeZerosDialog method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.ordering_metadata">ordering_metadata (pytransit.specific_tools.tnseq_tools.CombinedWigMetadata property)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.orf">orf (pytransit.specific_tools.tnseq_tools.Gene attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.orf2index">orf2index (pytransit.specific_tools.tnseq_tools.Genes attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="P">P</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.parse_lib_index">parse_lib_index() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile.phase_index">phase_index (pytransit.specific_tools.tnseq_tools.GffFile attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.phi">phi() (pytransit.specific_tools.tnseq_tools.Gene method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.phi_coefficient">phi_coefficient() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="pytransit.html#pytransit.transit_gui.PngFile">PngFile (class in pytransit.transit_gui)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.position">position (pytransit.specific_tools.tnseq_tools.Gene attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.PositionsAndReads">PositionsAndReads (pytransit.specific_tools.tnseq_tools.CombinedWig attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.protTable">protTable (pytransit.specific_tools.tnseq_tools.Genes attribute)</a>
</li>
      <li>
    pytransit

      <ul>
        <li><a href="pytransit.html#module-pytransit">module</a>, <a href="transit.html#module-pytransit">[1]</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    pytransit.specific_tools.norm_tools

      <ul>
        <li><a href="transit.html#module-pytransit.specific_tools.norm_tools">module</a>
</li>
      </ul></li>
      <li>
    pytransit.specific_tools.stat_tools

      <ul>
        <li><a href="transit.html#module-pytransit.specific_tools.stat_tools">module</a>
</li>
      </ul></li>
      <li>
    pytransit.specific_tools.tnseq_tools

      <ul>
        <li><a href="transit.html#module-pytransit.specific_tools.tnseq_tools">module</a>
</li>
      </ul></li>
      <li>
    pytransit.specific_tools.transit_tools

      <ul>
        <li><a href="transit.html#module-pytransit.specific_tools.transit_tools">module</a>
</li>
      </ul></li>
      <li>
    pytransit.transit_gui

      <ul>
        <li><a href="pytransit.html#module-pytransit.transit_gui">module</a>
</li>
      </ul></li>
  </ul></td>
</tr></table>

<h2 id="Q">Q</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.q_bernoulli_diff_distribution">q_bernoulli_diff_distribution() (in module pytransit.specific_tools.stat_tools)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.q_binomial_diff_distribution">q_binomial_diff_distribution() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.QuantileNorm">QuantileNorm (class in pytransit.specific_tools.norm_tools)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="R">R</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.read_condition_data">read_condition_data() (pytransit.specific_tools.tnseq_tools.CombinedWigMetadata static method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.read_counts_array">read_counts_array (pytransit.specific_tools.tnseq_tools.CombinedWig property)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.read_counts_by_wig_fingerprint">read_counts_by_wig_fingerprint (pytransit.specific_tools.tnseq_tools.CombinedWig property)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.read_genome">read_genome() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.read_result">read_result() (in module pytransit.specific_tools.tnseq_tools)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.transit_tools.read_result">(in module pytransit.specific_tools.transit_tools)</a>
</li>
      </ul></li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.read_results_file">read_results_file() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.reads">reads (pytransit.specific_tools.tnseq_tools.Gene attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.regress">regress() (in module pytransit.specific_tools.stat_tools)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.reps">reps (pytransit.specific_tools.tnseq_tools.Genes attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.require_r_to_be_installed">require_r_to_be_installed() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.resampling">resampling() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.ResultsFile">ResultsFile() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.row_for">row_for() (pytransit.specific_tools.tnseq_tools.CombinedWigMetadata method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.run_index">run_index() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.runs">runs() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.runs_w_info">runs_w_info() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.rv_site_indexes_map">rv_site_indexes_map() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="S">S</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.sample_trunc_norm_post">sample_trunc_norm_post() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile.score_index">score_index (pytransit.specific_tools.tnseq_tools.GffFile attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.select_conditions">select_conditions() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Wig.selected_as_gathered_data">selected_as_gathered_data() (pytransit.specific_tools.tnseq_tools.Wig static method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile.seqid_index">seqid_index (pytransit.specific_tools.tnseq_tools.GffFile attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.site_restricted_permutation">site_restricted_permutation() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile.source_index">source_index (pytransit.specific_tools.tnseq_tools.GffFile attribute)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.special_headers">special_headers (pytransit.specific_tools.tnseq_tools.CombinedWigMetadata attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.start">start (pytransit.specific_tools.tnseq_tools.Gene attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile.start_index">start_index (pytransit.specific_tools.tnseq_tools.GffFile attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.stats_for_gene">stats_for_gene() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.strand">strand (pytransit.specific_tools.tnseq_tools.Gene attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile.strand_index">strand_index (pytransit.specific_tools.tnseq_tools.GffFile attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.summed">summed() (pytransit.specific_tools.tnseq_tools.CombinedWig method)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="T">T</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.ta_sites">ta_sites (pytransit.specific_tools.tnseq_tools.CombinedWig property)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.text_histogram">text_histogram() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.theta">theta() (pytransit.specific_tools.tnseq_tools.Gene method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.TimerAndOutputs">TimerAndOutputs (class in pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="pytransit.html#pytransit.transit_gui.TnSeqFrame">TnSeqFrame (class in pytransit.transit_gui)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.tosses">tosses() (pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.tossify">tossify() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Gene.total_reads">total_reads() (pytransit.specific_tools.tnseq_tools.Gene method)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.total_reads">(pytransit.specific_tools.tnseq_tools.Genes method)</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.TotReadsNorm">TotReadsNorm (class in pytransit.specific_tools.norm_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.tri_coeff">tri_coeff() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.stat_tools.tri_cube">tri_cube() (in module pytransit.specific_tools.stat_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.trimmed_empirical_mu">trimmed_empirical_mu() (in module pytransit.specific_tools.norm_tools)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.TTRNorm.trimmed_empirical_mu">(pytransit.specific_tools.norm_tools.TTRNorm method)</a>
</li>
      </ul></li>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.TTRNorm">TTRNorm (class in pytransit.specific_tools.norm_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.GffFile.type_index">type_index (pytransit.specific_tools.tnseq_tools.GffFile attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="V">V</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.validate_wig_format">validate_wig_format() (in module pytransit.specific_tools.transit_tools)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.variance_run">variance_run() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="W">W</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Wig">Wig (class in pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.wig_fingerprints">wig_fingerprints (pytransit.specific_tools.tnseq_tools.CombinedWig property)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.wig_fingerprints">(pytransit.specific_tools.tnseq_tools.CombinedWigMetadata property)</a>
</li>
      </ul></li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.wig_id_suggestions_from_filepaths">wig_id_suggestions_from_filepaths() (in module pytransit.specific_tools.tnseq_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.wig_ids">wig_ids (pytransit.specific_tools.tnseq_tools.CombinedWig property)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.wig_ids">(pytransit.specific_tools.tnseq_tools.CombinedWigMetadata property)</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.Genes.wig_list">wig_list (pytransit.specific_tools.tnseq_tools.Genes attribute)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.wig_with_id">wig_with_id() (pytransit.specific_tools.tnseq_tools.CombinedWig method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.winsorize">winsorize() (in module pytransit.specific_tools.transit_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.with_loess_correction">with_loess_correction() (pytransit.specific_tools.tnseq_tools.CombinedWig method)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWig.with_only">with_only() (pytransit.specific_tools.tnseq_tools.CombinedWig method)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.tnseq_tools.CombinedWigMetadata.with_only">(pytransit.specific_tools.tnseq_tools.CombinedWigMetadata method)</a>
</li>
      </ul></li>
      <li><a href="transit.html#pytransit.specific_tools.transit_tools.write_result">write_result() (in module pytransit.specific_tools.transit_tools)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="Z">Z</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.zero_inflated_nb_objective_function">zero_inflated_nb_objective_function() (in module pytransit.specific_tools.norm_tools)</a>

      <ul>
        <li><a href="transit.html#pytransit.specific_tools.norm_tools.EmpHistNorm.zero_inflated_nb_objective_function">(pytransit.specific_tools.norm_tools.EmpHistNorm static method)</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.ZeroInflatedNBNorm">ZeroInflatedNBNorm (class in pytransit.specific_tools.norm_tools)</a>
</li>
      <li><a href="transit.html#pytransit.specific_tools.norm_tools.zinfnb_factors">zinfnb_factors() (in module pytransit.specific_tools.norm_tools)</a>
</li>
  </ul></td>
</tr></table>



           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2015, Michael A. DeJesus &amp; Thomas R. Ioerger.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(false);
      });
  </script>

  
  
    
   

</body>
</html>